sparse/tests/test_elemwise.py

```diff
@@ -0,0 +1,707 @@
+        ((1, 4), (3, 4)),
+        ((2, 2, 2), (1, 1, 1)),
+    ],
+)
+@pytest.mark.parametrize("format", [COO, GCXS])
+def test_elemwise_mixed(shape1, shape2, format):
+    s1 = sparse.random(shape1, density=0.5, format=format)
+    x2 = np.random.rand(*shape2)
+
+    x1 = s1.todense()
+
+    assert_eq(s1 * x2, x1 * x2)
+
+
+def test_elemwise_mixed_empty():
+    s1 = sparse.random((2, 0, 4), density=0.5)
+    x2 = np.random.rand(2, 0, 4)
+
+    x1 = s1.todense()
+
+    assert_eq(s1 * x2, x1 * x2)
+
+
+def test_elemwise_unsupported():
+    class A:
+        pass
+
+    s1 = sparse.random((2, 3, 4), density=0.5)
+    x2 = A()
+
+    with pytest.raises(TypeError):
+        s1 + x2
+
+    assert sparse.elemwise(operator.add, s1, x2) is NotImplemented
+
+
+def test_elemwise_mixed_broadcast():
+    s1 = sparse.random((2, 3, 4), density=0.5)
+    s2 = sparse.random(4, density=0.5)
+    x3 = np.random.rand(3, 4)
+
+    x1 = s1.todense()
+    x2 = s2.todense()
+
+    def func(x1, x2, x3):
+        return x1 * x2 * x3
+
+    assert_eq(sparse.elemwise(func, s1, s2, x3), func(x1, x2, x3))
+
+
+@pytest.mark.parametrize(
+    "func",
+    [operator.mul, operator.add, operator.sub, operator.gt, operator.lt, operator.ne],
+)
+@pytest.mark.parametrize("shape", [(2,), (2, 3), (2, 3, 4), (2, 3, 4, 5)])
+@pytest.mark.parametrize("format", [COO, GCXS])
+def test_elemwise_binary(func, shape, format):
+    xs = sparse.random(shape, density=0.5, format=format)
+    ys = sparse.random(shape, density=0.5, format=format)
+
+    x = xs.todense()
+    y = ys.todense()
+
+    assert_eq(func(xs, ys), func(x, y))
+
+
+@pytest.mark.parametrize("func", [operator.imul, operator.iadd, operator.isub])
+@pytest.mark.parametrize("shape", [(2,), (2, 3), (2, 3, 4), (2, 3, 4, 5)])
+@pytest.mark.parametrize("format", [COO, GCXS])
+def test_elemwise_binary_inplace(func, shape, format):
+    xs = sparse.random(shape, density=0.5, format=format)
+    ys = sparse.random(shape, density=0.5, format=format)
+
+    x = xs.todense()
+    y = ys.todense()
+
+    xs = func(xs, ys)
+    x = func(x, y)
+
+    assert_eq(xs, x)
+
+
+@pytest.mark.parametrize(
+    "func",
+    [
+        lambda x, y, z: x + y + z,
+        lambda x, y, z: x * y * z,
+        lambda x, y, z: x + y * z,
+        lambda x, y, z: (x + y) * z,
+    ],
+)
+@pytest.mark.parametrize("shape", [(2,), (2, 3), (2, 3, 4), (2, 3, 4, 5)])
+@pytest.mark.parametrize(
+    "formats",
+    [
+        [COO, COO, COO],
+        [GCXS, GCXS, GCXS],
+        [COO, GCXS, GCXS],
+    ],
+)
+def test_elemwise_trinary(func, shape, formats):
+    xs = sparse.random(shape, density=0.5, format=formats[0])
+    ys = sparse.random(shape, density=0.5, format=formats[1])
+    zs = sparse.random(shape, density=0.5, format=formats[2])
+
+    x = xs.todense()
+    y = ys.todense()
+    z = zs.todense()
+
+    fs = sparse.elemwise(func, xs, ys, zs)
+    if COO in formats:
+        assert isinstance(fs, COO)
+    else:
+        assert isinstance(fs, GCXS)
```

Comment:
I would remove any guarantees here about mixed formats. If the formats are the same, then, yes, please, test that `format_in == format_out`.